File: src/csharp/FlashAlpha.QuantConnect/Data/StockQuote.cs
using System;
using System.Text.Json.Serialization;
using QuantConnect.Data;

namespace FlashAlpha.QuantConnect.Data;

/// <summary>
/// Stock-quote bar (bid/ask/mid/last) from the FlashAlpha historical API.
/// </summary>
/// <remarks>
/// <para>Mirrors the JSON response from <c>GET /v1/stockquote/{symbol}?at=...</c>.
/// The historical SDK's <c>StockQuoteAsync</c> returns a raw
/// <see cref="System.Text.Json.JsonElement"/> (no typed model in the SDK), so
/// the bar carries each documented field explicitly.</para>
///
/// <para><b>Wire-format note:</b> the upstream feed uses camelCase for
/// <c>lastUpdate</c> (NOT <c>last_update</c>) — preserved here with an explicit
/// <see cref="JsonPropertyNameAttribute"/>.</para>
///
/// <para>The bar root key is <c>ticker</c> (not <c>symbol</c>) on this endpoint
/// — see <see cref="Ticker"/>'s attribute.</para>
///
/// <para>Subscribe with: <c>algo.AddData&lt;FlashAlphaStockQuoteBar&gt;("SPY", Resolution.Daily);</c>
/// or the sugar <see cref="QCAlgorithmExtensions.AddFlashAlphaStockQuote"/>.</para>
/// </remarks>
public class FlashAlphaStockQuoteBar : BaseData
{
    /// <summary>Ticker echoed by the API. Symbol on <see cref="BaseData"/> is the LEAN identity.</summary>
    [JsonPropertyName("ticker")]
    public string? Ticker { get; set; }

    /// <summary>Best bid for the underlier at <see cref="LastUpdate"/>.</summary>
    [JsonPropertyName("bid")]
    public double? Bid { get; set; }

    /// <summary>Best ask for the underlier at <see cref="LastUpdate"/>.</summary>
    [JsonPropertyName("ask")]
    public double? Ask { get; set; }

    /// <summary>Mid price: <c>(bid + ask) / 2</c>.</summary>
    [JsonPropertyName("mid")]
    public double? Mid { get; set; }

    /// <summary>Last trade price for the underlier.</summary>
    [JsonPropertyName("last")]
    public double? Last { get; set; }

    /// <summary>Last quote/trade update timestamp. Wire field is camelCase.</summary>
    [JsonPropertyName("lastUpdate")]
    public string? LastUpdate { get; set; }

    /// <inheritdoc cref="FlashAlphaGexBar.GetSource"/>
    public override SubscriptionDataSource GetSource(
        SubscriptionDataConfig config, DateTime date, bool isLiveMode)
        => FlashAlphaSource.For("stock/quote", config.Symbol, date);

    /// <inheritdoc cref="FlashAlphaGexBar.Reader"/>
    public override BaseData Reader(
        SubscriptionDataConfig config, string line, DateTime date, bool isLiveMode)
        => FlashAlphaSource.Parse<FlashAlphaStockQuoteBar>(line, config.Symbol, date) ?? new FlashAlphaStockQuoteBar();
}

File: src/python/src/flashalpha_quantconnect/data/stock_quote.py
"""Stock-quote bar (bid/ask/mid/last) from
``GET /v1/stockquote/{symbol}?at=...``.

The historical SDK's ``stock_quote()`` returns a plain ``dict`` (no
typed model), so this bar declares each documented field explicitly.

Wire-format note: the root key is ``ticker`` (NOT ``symbol``) and the
timestamp field is camelCase ``lastUpdate``. The ``_field_aliases``
declaration captures both.
"""

from __future__ import annotations

from typing import Any, Optional

from ._lean_import import PythonDataBase
from .source import source_for, parse


class StockQuoteBar(PythonDataBase):
    """Stock-quote bar — top-of-book bid/ask/mid/last for the underlier.

    Mirrors the JSON response from ``GET /v1/stockquote/{symbol}?at=...``.

    Subscribe with: ``algo.AddData(StockQuoteBar, ticker, Resolution.Daily)``
    or the sugar ``add_flashalpha_stock_quote(algo, "SPY")``.

    Invariants:
        ``bar.Bid <= bar.Mid <= bar.Ask`` (within feed precision).
    """

    Ticker: str = ""
    Bid: Optional[float] = None
    Ask: Optional[float] = None
    Mid: Optional[float] = None
    Last: Optional[float] = None
    LastUpdate: str = ""

    _field_aliases = {
        "Ticker": "ticker",
        "LastUpdate": "lastUpdate",
    }

    def GetSource(self, config: Any, date: Any, is_live_mode: bool) -> Any:
        return source_for("stock/quote", config.Symbol, date)

    def Reader(self, config: Any, line: str, date: Any, is_live_mode: bool) -> Any:
        return parse(StockQuoteBar, line, config.Symbol, date)
